### ISSUE TITLE
Rebase or reset doesn't work and locks up Sublime Merge

### ISSUE THREAD

**Issue by lgritz** [sublimehq/sublime_merge#1098]
**Version info**

 - OS: MacOS 10.15.7
 - Build: 2049

**Description**

I’ve used SM for a long time and just in the last couple days (perhaps starting when I upgraded to build 2049?), I’ve noticed that some fairly ordinary operations just hang. Resetting a branch to a different commit. Doing a simple rebase of a branch onto a different spot. You give the command and nothing seems to happen, and the branch marker doesn’t move. But process meter shows SM taking 100%.

It doesn’t happen for all resets and rebases, but it is common, and when it happens, it will happen consistently if I try to do the same reset/rebase again. For example, exiting SublimeMerge and restarting does not help, it will continue to fail on the same reset no matter how many times I try. (Though doing the same action from the command line or from gitk is fine.) I haven’t noticed a pattern that lets me predict if whether a particular reset or rebase will work or not. When it fails, SM will take 100% CPU until I exit. If I try to do other operations or do it again without an exit, it will lock up and needs a Force Quit.

**Steps to reproduce**

Often happens when want to simply move a branch marker to a different commit.

1. select a branch and check out
2. right clock on a different commit, select 'Reset to this Commit' -> 'Hard'

But nothing seems to happen, branch marker does not move to the other commit, no error message or dialog. But now the Sublime Merge process is chewing up 100% CPU.

**Expected behavior**

Confirmation dialog and branch marker resets to the desired commit.

**Debug information**

```
=== App Version Information ===
Build: 2049

=== Git Version Information ===
Using Git: git (system)
git version 2.31.0
PATH: /usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin:/opt/X11/bin:/Library/Apple/usr/bin (from shell)
environment variables loaded using: /bin/bash -l

=== Browse Page Information ===
HEAD: 867bd0bee27db23acaa420cf2abf356962e1aa2d
Is in merge: 0
Is in cherry_pick: 0
Is in rebase: 0
Is in revert: 0

=== Git Status Information ===


=== Our Status Information ===

=== Git Config Information ===

=Closes #$1."; 		fi 	}; f
=alias.add-unmerged=!f() { git diff --name-status --diff-filter=U | cut -f2 ; }; git add `f`
alias.amend=!f() { git commit --amend $@; } ; f
alias.amenda=!f() { git commit --amend -a $@; } ; f
alias.amendax=!f() { git commit --amend --no-edit -a $@; } ; f
alias.amendx=!f() { git commit --amend --no-edit $@; } ; f
alias.branches=!legit branches
alias.credit=!f() { git commit --amend --author "$1 <$2>" -C HEAD; }; f
alias.deletemerged=!git branch --merged | grep -v '\*' | xargs -n 1 git branch -d
alias.edit-unmerged=!f() { git diff --name-status --diff-filter=U | cut -f2 ; }; subl `f`
alias.graft=!legit graft "$@"
alias.harvest=!legit harvest "$@"
alias.mergepr=!f() { 		declare currentBranch="$(git symbolic-ref --short HEAD)"; 		declare branch="${2:-$currentBranch}"; 		if [ $(printf "%s" "$1" | grep '^[0-9]\+$' > /dev/null; printf $?) -eq 0 ]; then 			git fetch origin refs/pull/$1/head:pr/$1 && 			git checkout -B $branch && 			git rebase $branch pr/$1 && 			git checkout -B $branch && 			git merge pr/$1 && 			git branch -D pr/$1 && 			git commit --amend -m "$(git log -1 --pretty=%B)
alias.publish=!legit publish "$@"
alias.reb=!r() { git rebase -i HEAD~$1; }; r
alias.redate=!f() { git commit --amend --date "$1" -C HEAD; }; f
alias.redatenow=!f() { git commit --amend --date "$1" -C HEAD; }; f $(date "+%Y-%m-%d %H:%M:%S")
alias.retag=!r() { git tag -d $1 && git push origin :refs/tags/$1 && git tag $1; }; r
alias.rmbranch=!f() { git branch -D "$1" ; git push origin --delete "$1"; }; f
alias.rmtag=!r() { git tag -d $1 && git push origin :refs/tags/$1; }; r
alias.sprout=!legit sprout "$@"
alias.switch=!legit switch "$@"
alias.sync=!legit sync "$@"
alias.unpublish=!legit unpublish "$@"
blame.ignorerevsfile=.git-blame-ignore-revs
branch.dankamongmen/video-crash.merge=refs/heads/dankamongmen/video-crash
branch.dankamongmen/video-crash.remote=https://github.com/dankamongmen/oiio.git
color.diff.new=green
color.diff.old=red
core.autocrlf=input
core.bare=false
core.editor=emacs
core.filemode=true
core.logallrefupdates=true
core.repositoryformatversion=0
credential.helper=osxkeychain
diff.guitool=xxdiff
diff.tool=xxdiff
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.required=true
filter.lfs.smudge=git-lfs smudge -- %f
filter.media.clean=git media clean %f
filter.media.required=true
filter.media.smudge=git media smudge %f
gui.geometry=1037x812+368+27 194 203
gui.wmstate=normal
merge.tool=xxdiff
remote.OpenImageIO.fetch=+refs/heads/*:refs/remotes/OpenImageIO/*
remote.OpenImageIO.gh-resolved=base
remote.OpenImageIO.url=git@github.com:OpenImageIO/oiio.git
remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
remote.origin.url=git@github.com:lgritz/oiio.git
user.email=lg@larrygritz.com
user.name=Larry Gritz

=== Our Config Information ===
Git Config Path Information
Using config path: /etc/gitconfig
Using config path: /Users/lg/.config/git/config
Using config path: /Users/lg/.gitconfig
Using config path: /Users/lg/code/oiio/oiio.lg/.git/config

Our config output is different to the Git config output

=Closes #$1."; 		fi 	}; f
=alias.add-unmerged=!f() { git diff --name-status --diff-filter=U | cut -f2 ; }; git add `f`
alias.amend=!f() { git commit --amend $@; } ; f
alias.amenda=!f() { git commit --amend -a $@; } ; f
alias.amendax=!f() { git commit --amend --no-edit -a $@; } ; f
alias.amendx=!f() { git commit --amend --no-edit $@; } ; f
alias.branches=!legit branches
alias.credit=!f() { git commit --amend --author "$1 <$2>" -C HEAD; }; f
alias.deletemerged=!git branch --merged | grep -v '\*' | xargs -n 1 git branch -d
alias.edit-unmerged=!f() { git diff --name-status --diff-filter=U | cut -f2 ; }; subl `f`
alias.graft=!legit graft "$@"
alias.harvest=!legit harvest "$@"
alias.mergepr=!f() { 		declare currentBranch="$(git symbolic-ref --short HEAD)"; 		declare branch="${2:-$currentBranch}"; 		if [ $(printf "%s" "$1" | grep '^[0-9]\+$' > /dev/null; printf $?) -eq 0 ]; then 			git fetch origin refs/pull/$1/head:pr/$1 && 			git checkout -B $branch && 			git rebase $branch pr/$1 && 			git checkout -B $branch && 			git merge pr/$1 && 			git branch -D pr/$1 && 			git commit --amend -m "$(git log -1 --pretty=%B)
alias.publish=!legit publish "$@"
alias.reb=!r() { git rebase -i HEAD~$1; }; r
alias.redate=!f() { git commit --amend --date "$1" -C HEAD; }; f
alias.redatenow=!f() { git commit --amend --date "$1" -C HEAD; }; f $(date "+%Y-%m-%d %H:%M:%S")
alias.retag=!r() { git tag -d $1 && git push origin :refs/tags/$1 && git tag $1; }; r
alias.rmbranch=!f() { git branch -D "$1" ; git push origin --delete "$1"; }; f
alias.rmtag=!r() { git tag -d $1 && git push origin :refs/tags/$1; }; r
alias.sprout=!legit sprout "$@"
alias.switch=!legit switch "$@"
alias.sync=!legit sync "$@"
alias.unpublish=!legit unpublish "$@"
blame.ignorerevsfile=.git-blame-ignore-revs
branch.dankamongmen/video-crash.merge=refs/heads/dankamongmen/video-crash
branch.dankamongmen/video-crash.remote=https://github.com/dankamongmen/oiio.git
color.diff.new=green
color.diff.old=red
core.autocrlf=input
core.bare=false
core.editor=emacs
core.filemode=true
core.logallrefupdates=true
core.repositoryformatversion=0
diff.guitool=xxdiff
diff.tool=xxdiff
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.required=true
filter.lfs.smudge=git-lfs smudge -- %f
filter.media.clean=git media clean %f
filter.media.required=true
filter.media.smudge=git media smudge %f
gui.geometry=1037x812+368+27 194 203
gui.wmstate=normal
merge.tool=xxdiff
remote.OpenImageIO.fetch=+refs/heads/*:refs/remotes/OpenImageIO/*
remote.OpenImageIO.gh-resolved=base
remote.OpenImageIO.url=git@github.com:OpenImageIO/oiio.git
remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
remote.origin.url=git@github.com:lgritz/oiio.git
user.email=lg@larrygritz.com
user.name=Larry Gritz

=== Git Attributes Information ===
git check_attr --all output

=== Our Modified Files Newline Normalisation and EOL Information ===

=== Our Modified Files Flag Information ===
Ignoring symlinks: 0

```

**Comment by srbs** [comment]
It looks like you're experiencing this in the https://github.com/OpenImageIO/oiio repository. If the commits are pushed, can you specify the two commit hashes you are reseting from & to? (or possibly create temporary branches for reproduction)

I have not been able to reproduce so far on macOS 10.14.6.

**Comment by lgritz** [comment]
I didn't get to do much programming today, but as soon as I run across another action that hangs, I'll relay to you.

**Comment by lgritz** [comment]
OK, finally stumbled across a failure.

If you are synchronized to the oiio repository as it currently is,

1. Checkout 894e4667 (where the RB-2.2 branch is at this moment, but may not stay there depending on when you sync).
2. Cherry-pick 91e9852ac2 (where master is at this moment, but may not stay there).

I select cherry-pick, but nothing happens. But now the fan is running and Sublime Merge is using 100% doing... hard to tell what, exactly.

I hope you can repro this. It's hard to tell how specific it may be to my local repository and whether you can recreate it with a fresh clone.

**Comment by lgritz** [comment]
Hmmm... after exiting Sublime Merge (after doing NOTHING after the failed step above), then re-launching SM, I was able to do the cherry-pick. Between the two attempts, I didn't do anything from the command line or touch the repo in any other way. So I'm not super confident that you can recreate it with those specific commits on your end, it does seems to be something about Sublime Merge's internal state that plays a factor.

But note that this is one of the few times that existing and restarting SM has resulted in progress. Usually when I've encountered this, no matter how many times I exit/restart, the problem persists.

**Comment by lgritz** [comment]
Thinking that maybe there was something corrupted or peculiar to my local repo, I tried creating a totally fresh clone of the OIIO repo and working out of there for a while. In fairly short order, I had the same freezes.

So "good" news is that maybe you will be able to reproduce the problem with a clone on your end and that it's not specific to something in my local repo?

But "bad" news is that it's not going to be as easy a workaround as just checking out a fresh repo.

Today I'm working with the SM "console" open that shows the underlying commands, I had not thought to do that before, so perhaps that will reveal something interesting the next time I encounter the problem. I will keep you posted.

**Comment by lgritz** [comment]
I got results rather quickly!

So, when the console is open, you issue a command through the GUI and the console will display (here using an example of selecting a new branch to checkout):

    Executing: git checkout master
    Working dir: /Users/lg/code/oiio/oiio.lg
      Env vars: SHELL=/bin/bash TMPDIR=/var/folders/92/9d0zy2n11_9cnmgmq65p91t00000gn/T/ USER=lg SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.tUcNTyfnQM/Listeners __CF_USER_TEXT_ENCODING=0x1F5:0x0:0x52 PATH=/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin:/opt/X11/bin:/Library/Apple/usr/bin XPC_FLAGS=0x0 XPC_SERVICE_NAME=0 HOME=/Users/lg LOGNAME=lg DISPLAY=/private/tmp/com.apple.launchd.gqBtcHjtUC/org.macosforge.xquartz:0 GIT_EDITOR=/usr/bin/true GIT_PAGER=cat
    Result: success
    head changed to 11cae6488033c931a9ff729403ca250e8e7e13ba

So the next thing I did was try to more the branch marker to a different commit with "Reset master to this commit -> Hard".

And it does the usual freeze.

The interesting thing is, no output went to the console from that operation at all. So whatever is going wrong happens before any console output. I hope maybe that helps you in some way narrow down when the problem might be occurring.

**Comment by lgritz** [comment]
I'm still experiencing this problem, and a new wrinkle has cropped up (maybe this is new, or maybe it has been doing it throughout this period of trouble and I hadn't noticed):

I'm quite sure that sometimes, I'll have a Sublime Merge window working, everything is fine, and I'll do some operation from the command line (a rebase, say), and then that SM window will peg the CPU meter at 100% and be in the kind of locked state I described.

In other words, just from the usual "watching" of the filesystem and updating itself when things change in my github local repo, that can jam Sublime Merge, in exactly the same way as doing certain commands in SM itself.
